### ISSUE TITLE
5.1.3 - RangeError: Maximum call stack size exceeded at inferFromTypes - Maybe Fixed with 5.2.0-dev.20230606

### ISSUE THREAD

**Issue by jpwogaman** [microsoft/TypeScript#54549]
# Bug Report

### 🔎 Search Terms

- "RangeError: Maximum call stack size exceeded at inferFromTypes"
- "inferFromTypes"

### 🕗 Version & Regression Information

- This changed between versions 5.0.4 and 5.1.3
- Seems to be fixed when running `npm install typescript@next //(5.2.0-dev.20230606)`

### 💻 Code

```cli
npm run build
```

### 🙁 Actual behavior

```cli
Build error occurred
RangeError: Maximum call stack size exceeded
    at inferFromTypes (C:\***\node_modules\typescript\lib\typescript.js:654
    at inferFromTypes (C:\***\node_modules\typescript\lib\typescript.js:65459:11)
    at inferFromTypes (C:\***\node_modules\typescript\lib\typescript.js:65459:11)
    at inferFromTypes (C:\***\node_modules\typescript\lib\typescript.js:65459:11)
    at inferFromTypes (C:\***\node_modules\typescript\lib\typescript.js:65459:11)
    at inferFromTypes (C:\***\node_modules\typescript\lib\typescript.js:65459:11)
    at inferFromTypes (C:\***\node_modules\typescript\lib\typescript.js:65459:11)
    at inferFromTypes (C:\***\node_modules\typescript\lib\typescript.js:654
    at inferFromTypes (C:\***\node_modules\typescript\lib\typescript.js:65459:11)
    at inferFromTypes (C:\***\node_modules\typescript\lib\typescript.js:65459:11) {
  type: 'RangeError'
}
```

### 🙂 Expected behavior

Build completed with no errors.

**Comment by dbrxnds** [comment]
Hi, we are also running into this issue on 5.1.3. Initially thought it was due to the size of our Prisma schema / types. 

Not sure what I can do to help debug this, but willing to if anyone has ideas.

**Comment by nicokunz** [comment]
We have this issue. too. Like @jpwogaman this is fixed for us by using a higher dev version of typescript:

`npm install typescript@next //(5.2.0-dev.20230606)`

**Comment by fatcerberus** [comment]
Nobody has a repro?

**Comment by AntQwanlity** [comment]
https://github.com/microsoft/TypeScript/issues/54348

**Comment by janpio** [comment]
Here is another potentially related case: https://github.com/microsoft/TypeScript/issues/54618
And I just commented there with a reproduction (Using Next.js, Prisma and Mongoose): https://github.com/microsoft/TypeScript/issues/54618#issuecomment-1590802825

**Comment by Jolg42** [comment]
With @janpio we could find that `5.2.0-dev.20230606` fixes the issue in mentioned reproduction.

I checked the changes that happened in 5.2 for that daily version I could see that 
https://github.com/microsoft/TypeScript/pull/54507 was backported to 5.1
https://github.com/microsoft/TypeScript/pull/54538 was not backported yet

So it seems to me that this would be fixed once https://github.com/microsoft/TypeScript/pull/54538 would be backported to 5.1?

**Comment by jakebailey** [comment]
The first PR should be enough to fix the crash. The second is a cleanup and potential perf boost (which we rarely backport).

This issue is really just a duplicate of #54348 which will be fixed in the next 5.1 patch (not yet released).

**Comment by suzan-rana** [comment]
Hi facing the same error using Next 13 app dir, prisma, typescript: 5.1.3, prisma: 4.15.0
![image](https://github.com/microsoft/TypeScript/assets/110159734/05feff50-f344-474a-825d-0987decd0316)
![image](https://github.com/microsoft/TypeScript/assets/110159734/db6ac082-cdfa-4d35-ad04-4b9366988632)

**Comment by suzan-rana** [comment]
It worked guys. 
![image](https://github.com/microsoft/TypeScript/assets/110159734/e89f2e45-f12b-4bde-9157-40b0b1e166f4)

**Comment by zakiego** [comment]
Also fixed with `5.2.0-dev.20230629`
